File: ts-src/cost/pricing.ts
export interface ModelPricingInfo {
	id: string;
	name: string;
	promptPrice: number;
	completionPrice: number;
	contextLength: number;
	maxCompletionTokens: number;
	modality: string;
	supportedParameters: string[];
}

interface ApiModelEntry {
	id: string;
	name: string;
	pricing: { prompt: string; completion: string };
	context_length: number;
	top_provider: { max_completion_tokens: number };
	architecture: { modality: string };
	supported_parameters: string[];
}

export class ModelPricing {
	private models: Map<string, ModelPricingInfo> | null = null;
	private fetchPromise: Promise<void> | null = null;
	private apiKey: string;
	private baseUrl: string;

	constructor(apiKey: string, baseUrl?: string) {
		this.apiKey = apiKey;
		this.baseUrl = baseUrl ?? "https://openrouter.ai/api/v1";
	}

	async getModel(modelId: string): Promise<ModelPricingInfo | undefined> {
		await this.ensureLoaded();
		return this.models?.get(modelId);
	}

	async getAllModels(): Promise<ModelPricingInfo[]> {
		await this.ensureLoaded();
		return Array.from(this.models?.values() ?? []);
	}

	async estimateCost(modelId: string, promptTokens: number, completionTokens: number): Promise<number | null> {
		const model = await this.getModel(modelId);
		if (!model) return null;
		return promptTokens * model.promptPrice + completionTokens * model.completionPrice;
	}

	get isLoaded(): boolean {
		return this.models !== null;
	}

	private async ensureLoaded(): Promise<void> {
		if (this.models) return;
		if (this.fetchPromise) {
			await this.fetchPromise;
			return;
		}
		this.fetchPromise = this.fetchModels();
		await this.fetchPromise;
	}

	private async fetchModels(): Promise<void> {
		const res = await fetch(`${this.baseUrl}/models`, {
			headers: { Authorization: `Bearer ${this.apiKey}` },
		});
		if (!res.ok) {
			throw new Error(`Models API returned ${res.status}`);
		}
		const json = (await res.json()) as { data?: ApiModelEntry[] };
		if (!Array.isArray(json.data)) {
			throw new Error(`Models API returned unexpected shape: ${JSON.stringify(json).slice(0, 200)}`);
		}
		const map = new Map<string, ModelPricingInfo>();
		for (const entry of json.data) {
			map.set(entry.id, {
				id: entry.id,
				name: entry.name,
				promptPrice: Number.parseFloat(entry.pricing.prompt),
				completionPrice: Number.parseFloat(entry.pricing.completion),
				contextLength: entry.context_length,
				maxCompletionTokens: entry.top_provider.max_completion_tokens,
				modality: entry.architecture.modality,
				supportedParameters: entry.supported_parameters,
			});
		}
		this.models = map;
	}
}

File: src/cost/pricing.rs
//! Model pricing data fetched from the OpenRouter `/models` endpoint.

use std::collections::HashMap;
use std::sync::Arc;

use anyhow::{anyhow, Result};
use serde::Deserialize;
use tokio::sync::Mutex;

#[derive(Debug, Clone)]
pub struct ModelPricingInfo {
    pub id: String,
    pub name: String,
    pub prompt_price: f64,
    pub completion_price: f64,
    pub context_length: u64,
    pub max_completion_tokens: u64,
    pub modality: String,
    pub supported_parameters: Vec<String>,
}

#[derive(Debug, Deserialize)]
struct ApiModelEntry {
    id: String,
    name: String,
    pricing: ApiPricing,
    context_length: u64,
    top_provider: ApiTopProvider,
    architecture: ApiArchitecture,
    #[serde(default)]
    supported_parameters: Vec<String>,
}

#[derive(Debug, Deserialize)]
struct ApiPricing {
    prompt: String,
    completion: String,
}

#[derive(Debug, Deserialize)]
struct ApiTopProvider {
    #[serde(default)]
    max_completion_tokens: u64,
}

#[derive(Debug, Deserialize)]
struct ApiArchitecture {
    modality: String,
}

#[derive(Debug, Deserialize)]
struct ApiModelsResponse {
    data: Vec<ApiModelEntry>,
}

#[derive(Clone)]
pub struct ModelPricing {
    inner: Arc<Mutex<Inner>>,
    api_key: String,
    base_url: String,
}

struct Inner {
    models: Option<HashMap<String, ModelPricingInfo>>,
}

impl ModelPricing {
    pub fn new(api_key: impl Into<String>, base_url: Option<&str>) -> Self {
        Self {
            inner: Arc::new(Mutex::new(Inner { models: None })),
            api_key: api_key.into(),
            base_url: base_url
                .unwrap_or("https://openrouter.ai/api/v1")
                .to_string(),
        }
    }

    pub async fn get_model(&self, model_id: &str) -> Option<ModelPricingInfo> {
        self.lookup_model(model_id).await.ok().flatten()
    }

    pub async fn lookup_model(&self, model_id: &str) -> Result<Option<ModelPricingInfo>> {
        self.ensure_loaded().await?;
        let guard = self.inner.lock().await;
        Ok(guard.models.as_ref().and_then(|m| m.get(model_id)).cloned())
    }

    pub async fn get_all_models(&self) -> Vec<ModelPricingInfo> {
        self.list_models().await.unwrap_or_default()
    }

    pub async fn list_models(&self) -> Result<Vec<ModelPricingInfo>> {
        self.ensure_loaded().await?;
        let guard = self.inner.lock().await;
        Ok(guard
            .models
            .as_ref()
            .map(|m| m.values().cloned().collect())
            .unwrap_or_default())
    }

    pub async fn search_models(&self, query: &str, limit: usize) -> Result<Vec<ModelPricingInfo>> {
        let query = query.trim().to_lowercase();
        let mut models = self.list_models().await?;
        if !query.is_empty() {
            models.retain(|m| {
                m.id.to_lowercase().contains(&query) || m.name.to_lowercase().contains(&query)
            });
        }
        models.sort_by(|a, b| a.id.cmp(&b.id));
        if limit > 0 {
            models.truncate(limit);
        }
        Ok(models)
    }

    pub async fn estimate_cost(
        &self,
        model_id: &str,
        prompt_tokens: u64,
        completion_tokens: u64,
    ) -> Option<f64> {
        let model = self.get_model(model_id).await?;
        Some(
            prompt_tokens as f64 * model.prompt_price
                + completion_tokens as f64 * model.completion_price,
        )
    }

    pub async fn is_loaded(&self) -> bool {
        self.inner.lock().await.models.is_some()
    }

    async fn ensure_loaded(&self) -> Result<()> {
        // Hold the lock across the fetch so concurrent callers dedupe to a
        // single network round-trip (mirrors TS in-flight Promise sharing).
        let mut guard = self.inner.lock().await;
        if guard.models.is_some() {
            return Ok(());
        }
        let map = self.fetch_models().await?;
        guard.models = Some(map);
        Ok(())
    }

    async fn fetch_models(&self) -> Result<HashMap<String, ModelPricingInfo>> {
        let client = reqwest::Client::new();
        let resp = client
            .get(format!("{}/models", self.base_url))
            .bearer_auth(&self.api_key)
            .send()
            .await?;
        if !resp.status().is_success() {
            return Err(anyhow!("Models API returned {}", resp.status()));
        }
        let parsed: ApiModelsResponse = resp.json().await?;
        let mut map = HashMap::new();
        for entry in parsed.data {
            map.insert(
                entry.id.clone(),
                ModelPricingInfo {
                    id: entry.id,
                    name: entry.name,
                    prompt_price: entry.pricing.prompt.parse().unwrap_or(0.0),
                    completion_price: entry.pricing.completion.parse().unwrap_or(0.0),
                    context_length: entry.context_length,
                    max_completion_tokens: entry.top_provider.max_completion_tokens,
                    modality: entry.architecture.modality,
                    supported_parameters: entry.supported_parameters,
                },
            );
        }
        Ok(map)
    }
}
